// @expo/config-plugins/build/ios/DeviceFamily.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDeviceFamily = exports.formatDeviceFamilies = exports.getDeviceFamilies = exports.getIsTabletOnly = exports.getSupportsTablet = exports.withDeviceFamily = void 0;
const ios_plugins_1 = require("../plugins/ios-plugins");
const WarningAggregator = __importStar(require("../utils/warnings"));
const withDeviceFamily = config => {
    return ios_plugins_1.withXcodeProject(config, async (config) => {
        config.modResults = await setDeviceFamily(config, {
            project: config.modResults,
        });
        return config;
    });
};
exports.withDeviceFamily = withDeviceFamily;
function getSupportsTablet(config) {
    var _a;
    return !!((_a = config.ios) === null || _a === void 0 ? void 0 : _a.supportsTablet);
}
exports.getSupportsTablet = getSupportsTablet;
function getIsTabletOnly(config) {
    var _a;
    return !!((_a = config === null || config === void 0 ? void 0 : config.ios) === null || _a === void 0 ? void 0 : _a.isTabletOnly);
}
exports.getIsTabletOnly = getIsTabletOnly;
function getDeviceFamilies(config) {
    var _a;
    const supportsTablet = getSupportsTablet(config);
    const isTabletOnly = getIsTabletOnly(config);
    if (isTabletOnly && ((_a = config.ios) === null || _a === void 0 ? void 0 : _a.supportsTablet) === false) {
        WarningAggregator.addWarningIOS('device-family', `Found contradictory values: \`{ ios: { isTabletOnly: true, supportsTablet: false } }\`. Using \`{ isTabletOnly: true }\`.`);
    }
    // 1 is iPhone, 2 is iPad
    if (isTabletOnly) {
        return [2];
    }
    else if (supportsTablet) {
        return [1, 2];
    }
    else {
        // is iPhone only
        return [1];
    }
}
exports.getDeviceFamilies = getDeviceFamilies;
/**
 * Wrapping the families in double quotes is the only way to set a value with a comma in it.
 *
 * @param deviceFamilies
 */
function formatDeviceFamilies(deviceFamilies) {
    return `"${deviceFamilies.join(',')}"`;
}
exports.formatDeviceFamilies = formatDeviceFamilies;
/**
 * Add to pbxproj under TARGETED_DEVICE_FAMILY
 */
function setDeviceFamily(config, { project }) {
    const deviceFamilies = formatDeviceFamilies(getDeviceFamilies(config));
    const configurations = project.pbxXCBuildConfigurationSection();
    // @ts-ignore
    for (const { buildSettings } of Object.values(configurations || {})) {
        // Guessing that this is the best way to emulate Xcode.
        // Using `project.addToBuildSettings` modifies too many targets.
        if (typeof (buildSettings === null || buildSettings === void 0 ? void 0 : buildSettings.PRODUCT_NAME) !== 'undefined') {
            buildSettings.TARGETED_DEVICE_FAMILY = deviceFamilies;
        }
    }
    return project;
}
exports.setDeviceFamily = setDeviceFamily;
