// @expo/config-plugins/build/plugins/expo-plugins.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withExpoLegacyPlugins = exports.getExpoLegacyPlugins = exports.withExpoVersionedSDKPlugins = exports.withExpoAndroidPlugins = exports.withExpoIOSPlugins = void 0;
const AndroidConfig = __importStar(require("../android"));
const IOSConfig = __importStar(require("../ios"));
const expo_ads_admob_1 = __importDefault(require("./unversioned/expo-ads-admob"));
const expo_apple_authentication_1 = __importDefault(require("./unversioned/expo-apple-authentication"));
const expo_branch_1 = __importDefault(require("./unversioned/expo-branch"));
const expo_contacts_1 = __importDefault(require("./unversioned/expo-contacts"));
const expo_document_picker_1 = __importDefault(require("./unversioned/expo-document-picker"));
const expo_facebook_1 = __importDefault(require("./unversioned/expo-facebook"));
const expo_notifications_1 = __importDefault(require("./unversioned/expo-notifications"));
const expo_splash_screen_1 = __importDefault(require("./unversioned/expo-splash-screen"));
const expo_updates_1 = __importDefault(require("./unversioned/expo-updates"));
const react_native_maps_1 = __importDefault(require("./unversioned/react-native-maps"));
const withPlugins_1 = require("./withPlugins");
const withStaticPlugin_1 = require("./withStaticPlugin");
/**
 * Config plugin to apply all of the custom Expo iOS config plugins we support by default.
 * TODO: In the future most of this should go into versioned packages like expo-facebook, expo-updates, etc...
 */
const withExpoIOSPlugins = (config, { bundleIdentifier }) => {
    // Set the bundle ID ahead of time.
    if (!config.ios)
        config.ios = {};
    config.ios.bundleIdentifier = bundleIdentifier;
    return withPlugins_1.withPlugins(config, [
        [IOSConfig.BundleIdentifier.withBundleIdentifier, { bundleIdentifier }],
        IOSConfig.Swift.withSwiftBridgingHeader,
        IOSConfig.Swift.withNoopSwiftFile,
        IOSConfig.Google.withGoogle,
        IOSConfig.Name.withDisplayName,
        IOSConfig.Orientation.withOrientation,
        IOSConfig.RequiresFullScreen.withRequiresFullScreen,
        IOSConfig.Scheme.withScheme,
        IOSConfig.UserInterfaceStyle.withUserInterfaceStyle,
        IOSConfig.UsesNonExemptEncryption.withUsesNonExemptEncryption,
        IOSConfig.Version.withBuildNumber,
        IOSConfig.Version.withVersion,
        IOSConfig.Google.withGoogleServicesFile,
        // Entitlements
        IOSConfig.Entitlements.withAssociatedDomains,
        // XcodeProject
        IOSConfig.DeviceFamily.withDeviceFamily,
        IOSConfig.Locales.withLocales,
        // Dangerous
        IOSConfig.Icons.withIcons,
    ]);
};
exports.withExpoIOSPlugins = withExpoIOSPlugins;
/**
 * Config plugin to apply all of the custom Expo Android config plugins we support by default.
 * TODO: In the future most of this should go into versioned packages like expo-facebook, expo-updates, etc...
 */
const withExpoAndroidPlugins = (config, props) => {
    // Set the package name ahead of time.
    if (!config.android)
        config.android = {};
    config.android.package = props.package;
    return withPlugins_1.withPlugins(config, [
        // settings.gradle
        AndroidConfig.Name.withNameSettingsGradle,
        // project build.gradle
        AndroidConfig.GoogleServices.withClassPath,
        // app/build.gradle
        AndroidConfig.GoogleServices.withApplyPlugin,
        AndroidConfig.Package.withPackageGradle,
        AndroidConfig.Version.withVersion,
        // AndroidManifest.xml
        AndroidConfig.Package.withPackageManifest,
        AndroidConfig.AllowBackup.withAllowBackup,
        // Note: The withAndroidIntentFilters plugin must appear before the withScheme
        // plugin or withScheme will override the output of withAndroidIntentFilters.
        AndroidConfig.IntentFilters.withAndroidIntentFilters,
        AndroidConfig.Scheme.withScheme,
        AndroidConfig.Orientation.withOrientation,
        AndroidConfig.Permissions.withPermissions,
        AndroidConfig.UserInterfaceStyle.withUiModeManifest,
        // MainActivity.*
        AndroidConfig.UserInterfaceStyle.withUiModeMainActivity,
        // strings.xml
        AndroidConfig.Name.withName,
        // Dangerous -- these plugins run in reverse order.
        AndroidConfig.GoogleServices.withGoogleServicesFile,
        // Modify colors.xml and styles.xml
        AndroidConfig.RootViewBackgroundColor.withRootViewBackgroundColor,
        AndroidConfig.NavigationBar.withNavigationBar,
        AndroidConfig.StatusBar.withStatusBar,
        AndroidConfig.PrimaryColor.withPrimaryColor,
        AndroidConfig.Icon.withIcons,
        // If we renamed the package, we should also move it around and rename it in source files
        // Added last to ensure this plugin runs first. Out of tree solutions will mistakenly resolve the package incorrectly otherwise.
        AndroidConfig.Package.withPackageRefactor,
    ]);
};
exports.withExpoAndroidPlugins = withExpoAndroidPlugins;
const withExpoVersionedSDKPlugins = (config, { expoUsername }) => {
    return withPlugins_1.withPlugins(config, [
        react_native_maps_1.default,
        expo_ads_admob_1.default,
        expo_apple_authentication_1.default,
        expo_contacts_1.default,
        expo_notifications_1.default,
        [expo_updates_1.default, { expoUsername }],
        expo_branch_1.default,
        expo_document_picker_1.default,
        expo_facebook_1.default,
        expo_splash_screen_1.default,
    ]);
};
exports.withExpoVersionedSDKPlugins = withExpoVersionedSDKPlugins;
function getExpoLegacyPlugins() {
    return expoLegacyPlugins;
}
exports.getExpoLegacyPlugins = getExpoLegacyPlugins;
// Expo managed packages that require extra update.
// These get applied automatically to create parity with expo build in eas build.
const expoLegacyPlugins = [
    'expo-app-auth',
    'expo-av',
    'expo-background-fetch',
    'expo-barcode-scanner',
    'expo-brightness',
    'expo-calendar',
    'expo-camera',
    'expo-dev-menu',
    'expo-dev-launcher',
    'expo-dev-client',
    'expo-image-picker',
    'expo-file-system',
    'expo-ads-facebook',
    'expo-location',
    'expo-media-library',
    'expo-screen-orientation',
    'expo-sensors',
    'expo-task-manager',
    'expo-local-authentication',
];
// Plugins that need to be automatically applied, but also get applied by expo-cli if the versioned plugin isn't available.
// These are split up because the user doesn't need to be prompted to setup these packages.
const expoManagedVersionedPlugins = [
    'expo-firebase-analytics',
    'expo-firebase-core',
    'expo-google-sign-in',
];
const withOptionalLegacyPlugins = (config, plugins) => {
    return plugins.reduce((prev, plugin) => {
        return withStaticPlugin_1.withStaticPlugin(prev, {
            // hide errors
            _isLegacyPlugin: true,
            plugin,
            // If a plugin doesn't exist, do nothing.
            fallback: config => config,
        });
    }, config);
};
function withExpoLegacyPlugins(config) {
    return withOptionalLegacyPlugins(config, [
        ...new Set(expoManagedVersionedPlugins.concat(expoLegacyPlugins)),
    ]);
}
exports.withExpoLegacyPlugins = withExpoLegacyPlugins;
