// @expo/config-plugins/build/android/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Version = exports.UserInterfaceStyle = exports.Updates = exports.Styles = exports.Strings = exports.StatusBar = exports.SplashScreen = exports.Scheme = exports.RootViewBackgroundColor = exports.Resources = exports.Properties = exports.PrimaryColor = exports.Permissions = exports.Paths = exports.Package = exports.Orientation = exports.Notifications = exports.NavigationBar = exports.Name = exports.IntentFilters = exports.Icon = exports.GoogleServices = exports.AdMob = exports.GoogleMapsApiKey = exports.Facebook = exports.Colors = exports.Branch = exports.Manifest = exports.EasBuild = exports.AllowBackup = void 0;
const AdMob = __importStar(require("./AdMob"));
exports.AdMob = AdMob;
const AllowBackup = __importStar(require("./AllowBackup"));
exports.AllowBackup = AllowBackup;
const Branch = __importStar(require("./Branch"));
exports.Branch = Branch;
const Colors = __importStar(require("./Colors"));
exports.Colors = Colors;
const EasBuild = __importStar(require("./EasBuild"));
exports.EasBuild = EasBuild;
const Facebook = __importStar(require("./Facebook"));
exports.Facebook = Facebook;
const GoogleMapsApiKey = __importStar(require("./GoogleMapsApiKey"));
exports.GoogleMapsApiKey = GoogleMapsApiKey;
const GoogleServices = __importStar(require("./GoogleServices"));
exports.GoogleServices = GoogleServices;
const Icon = __importStar(require("./Icon"));
exports.Icon = Icon;
const IntentFilters = __importStar(require("./IntentFilters"));
exports.IntentFilters = IntentFilters;
const Manifest = __importStar(require("./Manifest"));
exports.Manifest = Manifest;
const Name = __importStar(require("./Name"));
exports.Name = Name;
const NavigationBar = __importStar(require("./NavigationBar"));
exports.NavigationBar = NavigationBar;
const Notifications = __importStar(require("./Notifications"));
exports.Notifications = Notifications;
const Orientation = __importStar(require("./Orientation"));
exports.Orientation = Orientation;
const Package = __importStar(require("./Package"));
exports.Package = Package;
const Paths = __importStar(require("./Paths"));
exports.Paths = Paths;
const Permissions = __importStar(require("./Permissions"));
exports.Permissions = Permissions;
const PrimaryColor = __importStar(require("./PrimaryColor"));
exports.PrimaryColor = PrimaryColor;
const Properties = __importStar(require("./Properties"));
exports.Properties = Properties;
const Resources = __importStar(require("./Resources"));
exports.Resources = Resources;
const RootViewBackgroundColor = __importStar(require("./RootViewBackgroundColor"));
exports.RootViewBackgroundColor = RootViewBackgroundColor;
const Scheme = __importStar(require("./Scheme"));
exports.Scheme = Scheme;
const SplashScreen = __importStar(require("./SplashScreen"));
exports.SplashScreen = SplashScreen;
const StatusBar = __importStar(require("./StatusBar"));
exports.StatusBar = StatusBar;
const Strings = __importStar(require("./Strings"));
exports.Strings = Strings;
const Styles = __importStar(require("./Styles"));
exports.Styles = Styles;
const Updates = __importStar(require("./Updates"));
exports.Updates = Updates;
const UserInterfaceStyle = __importStar(require("./UserInterfaceStyle"));
exports.UserInterfaceStyle = UserInterfaceStyle;
const Version = __importStar(require("./Version"));
exports.Version = Version;
